pkg/spanconfig/spanconfigreconciler/reconciler.go

```diff
@@ -583,6 +593,133 @@ func (r *incrementalReconciler) reconcile(
 	)
 }
 
+var cleanupOrphanedPTSRecordsInterval = settings.RegisterDurationSetting(
+	settings.TenantWritable,
+	"changefeed.protected_timestamp.cleanup_orphans_interval",
+	"this is the minimum interval which the reconciler must wait before checking for orphaned PTS"+
+		" records. setting a value of 0 disabled checking for orphaned PTS records",
+	12*time.Hour,
+	settings.NonNegativeDuration,
+)
+
+var orphanedPTSRecordStepUp = settings.RegisterDurationSetting(
+	settings.TenantWritable,
+	"changefeed.protected_timestamp.orphaned_pts_stepup",
+	"if an orphaned PTS record is older than the current PTS record by this amount, then, instead of being"+
+		" deleted, its timestamp is stepped up by this amount",
+	4*24*time.Hour,
+	settings.PositiveDuration,
+)
+
+var jobTypesEnrolledInStalePTSManagement = map[jobspb.Type]struct{}{
+	// NB: changefeed jobs have very specific semantics related to PTS records
+	// which allow for the automated cleanup of PTS records below. Give the
+	// methods in which this map is used a read before adding a new job
+	// type.
+	jobspb.TypeChangefeed: {},
```

Comment:
I'm having second and third thoughts about changing reconciler in order
to handle orphaned PTS records this way.

The fact that this is opt in means that if/when new issues arise, we cannot handle those issues easily.  The reality (unfortunate as it maybe) is that it's a safe bet that some things will happen in the future and we need to have a good mechanism on dealing with those.  More on this below.

I think that a much smaller, changefeed specific targeted fix as a first step as in https://github.com/cockroachdb/cockroach/pull/103893 might be best.
Furthermore, I think it can be made simpler still by removing step up logic and just deleting those orphan PTS records.   Also note, that a changefeed specific fix is safer in that we know that the running changefeed job only looks at records owned by that changefeed job and it knows that records are orphaned -- and that the update (deletion of orphan PTS records) happens under main transaction that updates job progress (all of this is to say that we definitely know these deletions/updates are safe from the changefeed correctness perspective).  Such fix is okay to backport -- which is something we would want to do.

It is not clear at all *if* the step up logic is beneficial -- as pointed out by @dt.  Instead of scanning 1 range once (512MB) to get rid of garbage, we will wind up scanning it many times; it's not at all clear that doing that will help.  The reality is that KV should not fall over when lots of garbage happens (as pointed out by @dt this can happen when e.g. large IMPORT is aborted -- we certainly expect things to work fine).  @irfansharif is working on this issue in #82955

Concretely, I propose a smaller scale fix (changefeed specific) as described above.  I think that you should take this code, and with few modifications introduce a new `crdb_internal.protected_timestamp_records` virtual table:
`(pts_id, owner_id, type (job/schedule), created, age, targets)`.
Would be nice to indicate number of ranges protected by a record -- but that might be too expensive -- so maybe just an indication if it's a db or a span.
We could also consider performing job/schedule lookup to determine owner type (i.e. changefeed, scheduled backup, etc)
The logic to populate that table is basically the query you have below.
Virtual table is very nice in that I can write queries:

```
SELECT * FROM crdb_internal.protected_timestamp_records
ORDER BY owner_id, age DESC
```
I can quickly see old records; I can quickly see which records are orhpaned -- because I know that changefeed ought to have only 1, etc.
I can then simply *delete* from pts table.

Because deletion is a bit more involved -- i.e. we have 2 table0 -- , we
might want to add `crdb_internal.drop_pts_record` builtin.
IF we think that advancing PTS timestamp is still a good idea (@erikgrinaker , @dt ?) -- we can add `crdb_internal.update_pts_timestamp` funciton.

Thoughts?

Reply:
Starting with some introspection tools sounds like a good plan to me -- doing so doesn't preclude coming back to do something automatic later if we determine the issue is widespread / requires more than manual intervention using those new introspection tools.

Perhaps we should also have a hard maximum PTS age, as a backstop of last resort to catch cases where it has slipped through the cracks. Something like 90 days (which is absurdly long, but again, this is a backstop of last resort; we expect a per-job or per-schedule option to expire after 3 days or something to kick in long before it, so it is only there to limit the blast radius of bugs). Then any uncleaned up orphans -- undetected since they're not hurting anyone but hanging out all the same -- would be cleaned up at some point instead of living on in perpetuity.

Reply:
I had my doubts about doing this in the reconciler as well, so I'm okay with a changefeed-specific fix. The issue is that the changefeed-specific fix in https://github.com/miretskiy/cockroach/commit/52516730b11aa92e98820c77600c4bd29476d9d9 requires a changefeed to actively be running - in fact it requires the changefeed which orphaned the PTS records to be running. Doing the change there does not guarantee that we will clean up all the orphans. If we are going to not do the step-up anyways, we can just write a backportable migration to clean up all the orphans.

Reply:
> backportable migration 

?

Reply:
By "backportable migration", I meant "some background process that would automatically clean up the PTS records".

Reply:
> I had my doubts about doing this in the reconciler as well, so I'm okay with a changefeed-specific fix. The issue is that the changefeed-specific fix in [miretskiy@5251673](https://github.com/miretskiy/cockroach/commit/52516730b11aa92e98820c77600c4bd29476d9d9) requires a changefeed to actively be running - in fact it requires the changefeed which orphaned the PTS records to be running. Doing the change there does not guarantee that we will clean up all the orphans. If we are going to not do the step-up anyways, we can just write a backportable migration to clean up all the orphans.

I don't think we need to fix ALL problems at once.  If non-running changefeed remains paused and holds to PTS record -- I think we should let those be. Bottom line: changefeeds (and any job) should be monitored.  If they aren't, there is only that much we can do.

We now have max pts age setting for changefeed, we also have PR that adds a 5 day max behind age (and I want to backport that).  I think the remaining corner case can remain in the corner case realm.

Reply:
> I've suggested gradually trimming the keyspan instead. This has the advantage of cleaning up all the garbage in a range all at once, instead of scanning all ranges a bunch of times.

Can you explain this more? Trimming the range -- like truncation? I don't think that's what you mean.

Reply:
Pasting from the Slack thread:

> I've suggested elsewhere that we can gradually reduce the keyspan of the PTS. If we compare the span to range bounds, we can know roughly how much data (and thus the upper bound of garbage) is in a given subspan. We can thus limit the maximum rate of GC by reducing the span gradually, e.g. by chopping off 1 range at both ends every x minutes. The MVCC GC queue will look for eligible ranges every 10 minutes.
>
>Of course, if we're doing this for 1 million ranges then 2 ranges every 3 minutes will take 3 years, so maybe we need to chop off e.g. 1% every 30 minutes and hope the ranges are uniformly distributed across stores.

In other words, if we have a PTS keyspan [a-z), and we know it contains 1 TB of data, we can reduce the PTS keyspan to [b-z), wait 30 minutes, then reduce it to [c-z), wait 30 minutes, etc. Then we'll be GCing 35 GB every 30 minutes. We can know the data distribution from the range bounds and stats, and can tune how much we chop off and how often we do it.

Reply:
Come to think of it, we can know the garbage distribution from the stats too, so we could make this pretty efficient.

Reply:
> I don't think we need to fix ALL problems at once. If non-running changefeed remains paused and holds to PTS record -- I think we should let those be. Bottom line: changefeeds (and any job) should be monitored. If they aren't, there is only that much we can do.
> 
> We now have max pts age setting for changefeed, we also have PR that adds a 5 day max behind age (and I want to backport that). I think the remaining corner case can remain in the corner case realm.

Okay, I think I'm convinced that doing it in the changefeed is ideal.  The `ptreconcile.Reconciler` (thanks @arulajmani) will clean up the PTS records for cancelled jobs. Paused jobs can keep holding on to PTS records. Running changefeeds will delete their own orphans. I don't see any other cases.

Reply:
@miretskiy discussed this a bit. KV has an open [issue](https://github.com/cockroachdb/cockroach/issues/82955) targeted for [23.2](https://cockroachlabs.slack.com/archives/C0KB9Q03D/p1685470550164769) to add admission control to mvcc gc. This should make it so that PTS records can be deleted freely without worrying about uncorking a large sudden burden on the cluster. It makes sense to wait for these changes (If needed, I'm happy to help out), then just add a simple fix in cdc code to delete old PTS records. 

cc @erikgrinaker @irfansharif

Reply:
I believe we would just need to wait for a code change to be made for https://github.com/cockroachdb/cockroach/issues/82955, as opposed to it being released to customers. Then we could merge our change on the CDC side and backport both changes. The work done on the KV side so far looks backportable, but the remaining work may not be. I'm hoping @irfansharif can comment on that, also if it will be worked on soon.

Reply:
I think we should consider adding the cdc side fix now.  It's better to delete these orphan records sooner than later.
At least, IMO.